service/history/historyReplicatorV2.go

```diff
@@ -0,0 +1,488 @@
+			if err != nil || !doContinue {
+				return err
+			}
+			return r.applyNonStartEvents(ctx, context, mutableState, task)
+		case *shared.EntityNotExistsError:
+			// mutable state not created, proceed
+			return r.applyNonStartEventsMissingMutableState(ctx, context, task)
+		default:
+			// unable to get mutable state, return err so we can retry the task later
+			return err
+		}
+	}
+}
+
+func (r *historyReplicatorV2) applyStartEvents(ctx ctx.Context, context workflowExecutionContext,
+	task historyReplicationTask) (retError error) {
+
+	requestID := uuid.New() // requestID used for start workflow execution request.  This is not on the history event.
+	msBuilder := r.getNewMutableState(task.getVersion(), task.getLogger())
+	sBuilder := r.getNewStateBuilder(msBuilder, task.getLogger())
+
+	// directly use stateBuilder to apply events for other events(including continueAsNew)
+	_, _, _, err := sBuilder.applyEvents(
+		task.getDomainID(),
+		requestID,
+		task.getExecution(),
+		task.getEvents(),
+		task.getNewRunEvents(),
+		persistence.EventStoreVersionV2, // 3+ DC will only use event store version 2
+		persistence.EventStoreVersionV2, // 3+ DC will only use event store version 2
+	)
+	if err != nil {
+		return err
+	}
+
+	_, err = context.appendFirstBatchEventsForStandby(msBuilder, task.getEvents())
+	if err != nil {
+		return err
+	}
+
+	// workflow passive side logic should not generate any replication task
+	createReplicationTask := false
+	transferTasks := sBuilder.getTransferTasks()
+	timerTasks := sBuilder.getTimerTasks()
+	defer func() {
+		if retError == nil {
+			r.notify(task.getSourceCluster(), task.getEventTime(), transferTasks, timerTasks)
+		}
+	}()
+
+	// try to create the workflow execution
+	createMode := persistence.CreateWorkflowModeBrandNew
+	prevRunID := ""
+	prevLastWriteVersion := int64(0)
+	err = context.createWorkflowExecution(
+		msBuilder, task.getSourceCluster(), createReplicationTask, task.getEventTime(), transferTasks, timerTasks,
+		createMode, prevRunID, prevLastWriteVersion,
+	)
+	if err == nil {
+		return nil
+	}
+	if _, ok := err.(*persistence.WorkflowExecutionAlreadyStartedError); !ok {
+		return err
+	}
+
+	// we have WorkflowExecutionAlreadyStartedError
+	errExist := err.(*persistence.WorkflowExecutionAlreadyStartedError)
+	currentRunID := errExist.RunID
+	currentState := errExist.State
+	currentLastWriteVersion := errExist.LastWriteVersion
+
+	if currentRunID == task.getRunID() {
+		return nil
+	}
+
+	// current workflow is completed
+	if currentState == persistence.WorkflowStateCompleted {
+		// allow the application of workflow creation if currentLastWriteVersion > incomingVersion
+		// because this can be caused by missing replication events
+		// proceed to create workflow
+		createMode = persistence.CreateWorkflowModeWorkflowIDReuse
+		prevRunID = currentRunID
+		prevLastWriteVersion = currentLastWriteVersion
+		return context.createWorkflowExecution(
```

Comment:
If the workflow is already started why create workflow execution again?

Reply:
basically, the replication path is trying to be optimistic.

each time, when creating a workflow, we need to know either there is no current record of that workflow, or that workflow is closed, plus its run ID and last write version.

the overall logic is
1. try to create the workflow optimistically, i.e. assuming that there is no existing workflow in DB
2. either 1 goes through, or see a WorkflowExecutionAlreadyStartedError
3. if see WorkflowExecutionAlreadyStartedError, then use the info in the error to continue create the workflow